metadata-ingestion/src/datahub/ingestion/transformer/add_dataset_dataproduct.py

```diff
@@ -0,0 +1,141 @@
+import logging
+from typing import Callable, Dict, List, Optional, Union
+
+import pydantic
+
+from datahub.configuration.common import (
+    ConfigurationError,
+    KeyValuePattern,
+    TransformerSemanticsConfigModel,
+)
+from datahub.configuration.import_resolver import pydantic_resolve_key
+from datahub.emitter.mce_builder import Aspect
+from datahub.emitter.mcp import MetadataChangeProposalWrapper
+from datahub.ingestion.api.common import PipelineContext
+from datahub.ingestion.transformer.dataset_transformer import (
+    DatasetDataproductTransformer,
+)
+from datahub.metadata.schema_classes import MetadataChangeProposalClass
+from datahub.specific.dataproduct import DataProductPatchBuilder
+
+logger = logging.getLogger(__name__)
+
+
+class AddDatasetDataProductConfig(TransformerSemanticsConfigModel):
+    # dataset_urn -> data product urn
+    get_data_product_to_add: Callable[[str], Optional[Union[str, List[str]]]]
```

Comment:
this should only return a single urn.

you'll need to modify the lambdas below to make them convert the list to a single value

```suggestion
    get_data_product_to_add: Callable[[str], Optional[str]]
```